File: benchmarks/python/json_vs_toon.py
"""
LTP JSON vs TOON Benchmark (Python)

Compares payload sizes and encoding/decoding performance
for JSON vs TOON encoding of large arrays
"""

import json
import time
import statistics
from typing import List, Dict, Any


def generate_affect_logs(count: int) -> List[Dict[str, Any]]:
    """Generate test data: large array of similar objects"""
    import math
    import random
    
    return [
        {
            't': i + 1,
            'valence': math.sin(i / 10) * 0.5,
            'arousal': math.cos(i / 10) * 0.5,
            'energy': random.random() * 0.5 + 0.5,
            'focus': random.random() * 0.3 + 0.7,
            'timestamp': int(time.time() * 1000) + i * 1000,
        }
        for i in range(count)
    ]


def format_bytes(bytes_size: int) -> str:
    """Format bytes to human-readable format"""
    if bytes_size == 0:
        return '0 B'
    
    for unit in ['B', 'KB', 'MB']:
        if bytes_size < 1024.0:
            return f'{bytes_size:.2f} {unit}'
        bytes_size /= 1024.0
    
    return f'{bytes_size:.2f} MB'


def benchmark(name: str, fn, iterations: int = 100) -> Dict[str, float]:
    """Run a benchmark and return statistics"""
    # Warmup
    for _ in range(10):
        fn()
    
    # Benchmark
    times = []
    for _ in range(iterations):
        start = time.perf_counter()
        fn()
        end = time.perf_counter()
        times.append((end - start) * 1000)  # Convert to milliseconds
    
    return {
        'name': name,
        'avg': statistics.mean(times),
        'min': min(times),
        'max': max(times),
        'total': sum(times),
    }


def simple_toon_encode(data: List[Dict[str, Any]]) -> str:
    """
    Simplified TOON encoding (for benchmark purposes)
    This is NOT a production-ready implementation
    """
    if not data or not isinstance(data, list):
        return json.dumps(data)
    
    if not isinstance(data[0], dict):
        return json.dumps(data)
    
    fields = list(data[0].keys())
    header = f"rows[{len(data)}]{{{','.join(fields)}}}:"
    lines = [
        "  " + ",".join(str(row.get(f, "")) for f in fields)
        for row in data
    ]
    return "\n".join([header] + lines)


def run_benchmarks():
    """Run JSON vs TOON benchmarks"""
    print("=== LTP JSON vs TOON Benchmark (Python) ===\n")
    
    sizes = [10, 100, 1000, 10000]
    
    for size in sizes:
        print(f"\n--- Array size: {size} objects ---")
        
        data = generate_affect_logs(size)
        
        # JSON encoding
        json_string = json.dumps(data)
        json_size = len(json_string.encode('utf-8'))
        
        # TOON encoding
        toon_string = simple_toon_encode(data)
        toon_size = len(toon_string.encode('utf-8'))
        
        # Size comparison
        reduction = (1 - toon_size / json_size) * 100
        
        print(f"JSON size:  {format_bytes(json_size)}")
        print(f"TOON size:  {format_bytes(toon_size)}")
        print(f"Reduction:  {reduction:.2f}%")
        
        # Encoding performance
        json_encode_bench = benchmark(
            'JSON encode',
            lambda: json.dumps(data),
            1000
        )
        
        toon_encode_bench = benchmark(
            'TOON encode',
            lambda: simple_toon_encode(data),
            1000
        )
        
        print(f"\nEncoding performance:")
        print(f"  JSON: avg {json_encode_bench['avg']:.3f}ms, "
              f"min {json_encode_bench['min']:.3f}ms, "
              f"max {json_encode_bench['max']:.3f}ms")
        print(f"  TOON: avg {toon_encode_bench['avg']:.3f}ms, "
              f"min {toon_encode_bench['min']:.3f}ms, "
              f"max {toon_encode_bench['max']:.3f}ms")
        
        # Decoding performance
        json_decode_bench = benchmark(
            'JSON decode',
            lambda: json.loads(json_string),
            1000
        )
        
        print(f"\nDecoding performance:")
        print(f"  JSON: avg {json_decode_bench['avg']:.3f}ms, "
              f"min {json_decode_bench['min']:.3f}ms, "
              f"max {json_decode_bench['max']:.3f}ms")
    
    print("\n=== Benchmark Complete ===")
    print("\nNote: TOON provides significant size reduction for large arrays")
    print("      of similar objects, making it ideal for affect logs,")
    print("      event batches, and telemetry data.")


if __name__ == '__main__':
    run_benchmarks()

File: benchmarks/js/json-vs-toon.js
/**
 * LTP JSON vs TOON Benchmark
 * 
 * Compares payload sizes and encoding/decoding performance
 * for JSON vs TOON encoding of large arrays
 */

const { simpleToonCodec } = require('../../examples/shared/simpleToonCodec');

// Generate test data: large array of similar objects
function generateAffectLogs(count) {
  return Array.from({ length: count }, (_, i) => ({
    t: i + 1,
    valence: Math.sin(i / 10) * 0.5,
    arousal: Math.cos(i / 10) * 0.5,
    energy: Math.random() * 0.5 + 0.5,
    focus: Math.random() * 0.3 + 0.7,
    timestamp: Date.now() + i * 1000,
  }));
}

function benchmark(name, fn, iterations = 100) {
  const times = [];
  
  // Warmup
  for (let i = 0; i < 10; i++) {
    fn();
  }
  
  // Benchmark
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    const end = performance.now();
    times.push(end - start);
  }
  
  const avg = times.reduce((a, b) => a + b, 0) / times.length;
  const min = Math.min(...times);
  const max = Math.max(...times);
  
  return { name, avg, min, max, total: times.reduce((a, b) => a + b, 0) };
}

function formatBytes(bytes) {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return Math.round(bytes / Math.pow(k, i) * 100) / 100 + ' ' + sizes[i];
}

function runBenchmarks() {
  console.log('=== LTP JSON vs TOON Benchmark ===\n');
  
  const sizes = [10, 100, 1000, 10000];
  
  for (const size of sizes) {
    console.log(`\n--- Array size: ${size} objects ---`);
    
    const data = generateAffectLogs(size);
    
    // JSON encoding
    const jsonString = JSON.stringify(data);
    const jsonSize = new Blob([jsonString]).size;
    
    // TOON encoding
    const toonString = simpleToonCodec.encodeJsonToToon(data);
    const toonSize = new Blob([toonString]).size;
    
    // Size comparison
    const reduction = ((1 - toonSize / jsonSize) * 100).toFixed(2);
    
    console.log(`JSON size:  ${formatBytes(jsonSize)}`);
    console.log(`TOON size:  ${formatBytes(toonSize)}`);
    console.log(`Reduction:  ${reduction}%`);
    
    // Encoding performance
    const jsonEncodeBench = benchmark(
      'JSON encode',
      () => JSON.stringify(data),
      1000
    );
    
    const toonEncodeBench = benchmark(
      'TOON encode',
      () => simpleToonCodec.encodeJsonToToon(data),
      1000
    );
    
    console.log(`\nEncoding performance:`);
    console.log(`  JSON: avg ${jsonEncodeBench.avg.toFixed(3)}ms, min ${jsonEncodeBench.min.toFixed(3)}ms, max ${jsonEncodeBench.max.toFixed(3)}ms`);
    console.log(`  TOON: avg ${toonEncodeBench.avg.toFixed(3)}ms, min ${toonEncodeBench.min.toFixed(3)}ms, max ${toonEncodeBench.max.toFixed(3)}ms`);
    
    // Decoding performance (JSON only, TOON decode not fully implemented)
    const jsonDecodeBench = benchmark(
      'JSON decode',
      () => JSON.parse(jsonString),
      1000
    );
    
    console.log(`\nDecoding performance:`);
    console.log(`  JSON: avg ${jsonDecodeBench.avg.toFixed(3)}ms, min ${jsonDecodeBench.min.toFixed(3)}ms, max ${jsonDecodeBench.max.toFixed(3)}ms`);
  }
  
  console.log('\n=== Benchmark Complete ===');
  console.log('\nNote: TOON provides significant size reduction for large arrays');
  console.log('      of similar objects, making it ideal for affect logs,');
  console.log('      event batches, and telemetry data.');
}

// Run if executed directly
if (require.main === module) {
  runBenchmarks();
}

module.exports = { runBenchmarks, generateAffectLogs };

